File: CH7(Backtracking)/Necklace.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
vector<vector<int>> a;
set<int> chosen;
long long ans = 0;

void solve(int i){
    if(i == n){
        ans++;
        return;
    }

    for(int x : a[i]){
        if(!chosen.count(x)){
            chosen.insert(x);

            solve(i + 1);

            chosen.erase(x);
        }
    }
}

int main(){
    cin >> n;

    a.resize(n);

    for(int i = 0; i < n; i++){
        int len;
        cin >> len;

        for(int j = 0; j < len; j++){
            int x;
            cin >> x;
            a[i].push_back(x);
        }
    }

    solve(0);

    cout << ans;
}

File: CH7(Backtracking)/Necklace.py
n = int(input())

a = []

for _ in range(n):
    x = list(map(int, input().split()))
    a.append(x[1:])

chosen = set()
ans = 0

def solve(i):
    global ans

    if i == n:
        ans += 1
        return

    for x in a[i]:
        if x not in chosen:
            chosen.add(x)

            solve(i + 1)

            chosen.remove(x)

solve(0)

print(ans)